snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java

```diff
@@ -15,273 +15,777 @@
+			throw new SnowowlRuntimeException("Failed to export terminology content to RF2.", e);
+		} finally {
+			if (exportDirectory != null) {
+				FileUtils.deleteDirectory(exportDirectory.toFile());
+			}
+		}
+
+		return exportId;
+	}
+
+	private TreeSet<CodeSystemVersionEntry> getVisibleVersions(final CodeSystemEntry codeSystemEntry) {
+		final TreeSet<CodeSystemVersionEntry> visibleVersions = newTreeSet(EFFECTIVE_DATE_ORDERING);
+		collectVersionsToExport(visibleVersions, codeSystemEntry, referenceBranch);
+		return visibleVersions;
+	}
+
+	private void collectVersionsToExport(final Set<CodeSystemVersionEntry> versionsToExport, final CodeSystemEntry codeSystemEntry, final String cutoffPath) {
+		final Collection<CodeSystemVersionEntry> candidates = getCodeSystemVersions(codeSystemEntry.getShortName());
+		if (candidates.isEmpty()) {
+			return;
+		}
+
+		final String versionParentPath = candidates.stream()
+				.map(CodeSystemVersionEntry::getParentBranchPath)
+				.findFirst()
+				.get();
+
+		final Set<String> versionNames = candidates.stream()
+				.map(CodeSystemVersionEntry::getVersionId)
+				.collect(Collectors.toSet());
+
+		final Branches versionBranches = getBranches(versionParentPath, versionNames);
+		final Map<String, Branch> versionBranchesByName = FluentIterable.from(versionBranches)
+				.uniqueIndex(b -> b.name());
+
+		final Branch cutoffBranch = getBranch(cutoffPath);
+		final long cutoffBaseTimestamp = getCutoffBaseTimestamp(cutoffBranch, versionParentPath);
+
+		// Remove all code system versions which were created after the cut-off date, or don't have a corresponding branch 
+		candidates.removeIf(v -> !versionBranchesByName.containsKey(v.getVersionId())
+				|| versionBranchesByName.get(v.getVersionId()).baseTimestamp() > cutoffBaseTimestamp);
+
+		versionsToExport.addAll(candidates);
+
+		// Exit early if only an extension code system should be exported, or we are already at the "base" code system
+		if (extensionOnly || Strings.isNullOrEmpty(codeSystemEntry.getExtensionOf())) {
+			return;
+		}
+
+		// Otherwise, collect applicable versions using this code system's working path
+		final CodeSystemEntry extensionEnty = getCodeSystem(codeSystemEntry.getExtensionOf());
+		collectVersionsToExport(versionsToExport, extensionEnty, codeSystemEntry.getBranchPath());
+	}
+
+	private void filterVersions(final TreeSet<CodeSystemVersionEntry> versionsToExport) {
+		// Is there anything to filter?
+		if (versionsToExport.isEmpty()) {
+			return;
+		}
+
+		switch (releaseType) {
+			case DELTA:
+				// Delta export: keep versions that fall into the effective time range
+				versionsToExport.removeIf(v -> {
+					return false
+							|| (startEffectiveTime != null && v.getEffectiveDate() < startEffectiveTime.getTime())
+							|| (endEffectiveTime != null && v.getEffectiveDate() > endEffectiveTime.getTime());
+				});
+				break;
+			case FULL:
+				// Full export: nothing to remove, all versions should be exported
```

Comment:
I would add/keep the ability to filter by effective time in case of Delta and Full export types.

Reply:
If you filter by effective time, it is no longer a full export. On the other hand, a delta export with the effective time range left untouched will give you a full export 😄

Reply:
You're right! 👍